sa/model.go

```diff
@@ -450,6 +451,57 @@ type authzModel struct {
 	ValidationRecord []byte     `db:"validationRecord"`
 }
 
+// rehydrateHostPort mutates a validation record. If the URL in the validation
+// record cannot be parsed, an error will be returned. If the Hostname and Port
+// fields already exist in the validation record, they will be retained.
+// Otherwise, the Hostname and Port will be derived and set from the URL field
+// of the validation record.
+func rehydrateHostPort(vr *core.ValidationRecord) error {
+	if vr.URL == "" {
+		return fmt.Errorf("rehydrating validation record, URL field cannot be empty")
+	}
+
+	parsedUrl, err := url.Parse(vr.URL)
+	if err != nil {
+		return fmt.Errorf("parsing validation record URL %q: %w", vr.URL, err)
+	}
+
+	if vr.Hostname == "" {
+		hostname := parsedUrl.Hostname()
+		if hostname == "" {
+			return fmt.Errorf("hostname missing in URL %q", vr.URL)
+		}
+		vr.Hostname = hostname
+	}
+
+	if vr.Port == "" {
+		// CABF BRs section 1.6.1: Authorized Ports: One of the following ports: 80
+		// (http), 443 (https)
+		if parsedUrl.Port() == "" {
+			// If there is only a scheme, then we'll determine the appropriate port.
+			switch parsedUrl.Scheme {
+			case "https":
+				vr.Port = "443"
+			case "http":
+				vr.Port = "80"
+			default:
+				// This should never happen since the VA should have already
+				// checked the scheme.
+				return fmt.Errorf("unknown scheme %q in URL %q", parsedUrl.Scheme, vr.URL)
+			}
+		} else if parsedUrl.Port() != "80" && parsedUrl.Port() != "443" {
+			// If :80 or :443 were explicitly stated in the URL field
+			// e.g. '"url":"https://example.com:443"'
+			return fmt.Errorf("only ports 80/tcp and 443/tcp are allowed in URL %q", vr.URL)
+		} else {
+			// This should never happen.
```

Comment:
Why should this never happen? What's to stop us from having a validation record with a url of `https://example.com:443`? Saying this should never happen is equivalent to saying "the url should never have a port embedded in it", which may be true, but then why are we trying to parse a port out of the url at all?

Basically I think there's just a mismatch between what the code is doing here, and what the comments are documenting. Either we expect there never to be an explicit port, and we should error out if there ever is; or we think having a port is fine, and the comments should reflect that.

Reply:
Whoops, good catch. That was some failed copy/paste while figuring out this section.